akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -0,0 +1,211 @@
+      if (nextElem ne null) {
+        needWakeup = false
+        if (onEvent(nextElem)) tryProcessNext(firstAttempt = true)
+      } else {
+        needWakeup = true
+        // additional poll() to grab any elements that might missed the needWakeup
+        // and have been enqueued just after it
+        if (firstAttempt)
+          tryProcessNext(firstAttempt = false)
+      }
+    }
+
+    def isShuttingDown: Boolean = shuttingDown
+
+    // External API
+    def enqueue(ev: Event): Unit = {
+      queue.add(ev)
+      if (needWakeup) {
+        needWakeup = false
+        wakeupCallback.invoke(NotUsed)
+      }
+    }
+
+    override def postStop(): Unit = {
+      // First announce that we are shutting down. This will notify late-comers to not even put anything in the queue
+      shuttingDown = true
+      // Anybody that missed the announcement needs to be notified.
+      var event = queue.poll()
+      while (event ne null) {
+        event match {
+          case Register(_, demandCallback) ⇒ demandCallback.invoke(MergeHub.Cancel)
+          case _                           ⇒
+        }
+        event = queue.poll()
+      }
+
+      // Kill everyone else
+      val states = demands.values().iterator()
+      while (states.hasNext) {
+        states.next().close()
+      }
+    }
+  }
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Sink[T, NotUsed]) = {
+    // TODO: Move inside the merge logic
+    val idCounter = new AtomicLong()
+    val producerCount = new AtomicLong()
+
+    val logic: MergedSourceLogic = new MergedSourceLogic(shape, idCounter)
+
+    val sink = new GraphStage[SinkShape[T]] {
+      val in: Inlet[T] = Inlet("MergeHub.in")
+      override val shape: SinkShape[T] = SinkShape(in)
+
+      override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+        // Start from non-zero demand to avoid initial delays.
+        // The HUB will expect this behavior.
+        private var demand: Long = perProducerBufferSize
+        private val id = idCounter.getAndIncrement()
+
+        override def preStart(): Unit = {
+          if (!logic.isShuttingDown) {
+            logic.enqueue(Register(id, getAsyncCallback(onDemand)))
+
+            // At this point, we could be in the unfortunate situation that:
+            // - we missed the shutdown announcement and entered this arm of the if statement
+            // - *before* we enqueued our Register event, the Hub already finished looking at the queue
+            //   and is now dead, so we are never notified again.
+            // To safeguard against this, we MUST check the announcement again. This is enough:
+            // if the Hub is no longer looking at the queue, then it must be that isShuttingDown must be already true.
+            if (!logic.isShuttingDown) pullWithDemand()
+            else completeStage()
```

Comment:
It is GC-ed if no one retains the Sink GraphStage. Since Sinks are reusable, it is expected that people keep it around in a val for longer times.
As for null-ing, it can't be done here, because at this time the Hub Source might be emptying the queue to deregister late-comers.